### PR TITLE
change the inputs of metrics and the calling methods in tests

### PR DESCRIPTION
**Pull request by bugtig6351**

### CODE REVIEW

File: rageval/metrics/_answer_accuracy.py [hunk 1]
```diff
@@ -1,121 +1,122 @@
...
+        answer: str,
+        gt_answer: str
+    ) -> float:
+        """Evaluating the correctness of answer."""
+        return answer == gt_answer
+
+    def _compute_batch(
+        self,
+        predictions,
+        references
```
Comment (faneshion) [inline rageval/metrics/_answer_accuracy.py:116]: I suggest using parameters`pred_answers` and `ref_answers` for clarity, as some metrics may require three or more parameters in addition to `predictions` and `references`.

File: rageval/metrics/_answer_bleu.py [hunk 2]
```diff
@@ -1,168 +1,165 @@
...
+
+    def _clean_special_tokens(self, sentence: str, subword: str) -> str:
+        """Clean special word in sentence"""
+
+        sentence = sentence.strip()
+        if subword is not None:
+            sentence = re.sub(subword, "", sentence)
+        return sentence
+
+    def _compute_single(self, dataset: datasets.Dataset) -> List[float]:
```
Comment (faneshion) [inline rageval/metrics/_answer_bleu.py:122]: This function has not been modified yet!

File: rageval/metrics/_answer_bleu.py [hunk 3]
```diff
@@ -1,168 +1,165 @@
...
+        batch_size: int = None,
+    ) -> Tuple[float, Dataset]:
+        """Evaluate the dataset."""
+
+        bleu = datasets.load_metric("bleu")
+        pred_tokens = [self._clean_special_tokens(pred).split(' ') for pred in predictions]
+        references_tokens = [[self._clean_special_tokens(gt).split(' ') for gt in ref] for ref in references]
+        bleu_result = bleu.compute(predictions=pred_tokens, references=references_tokens)
+        bleu_score = bleu_result['bleu']
+        scores = self._compute_single(predictions, references)
```
Comment (faneshion) [inline rageval/metrics/_answer_bleu.py:156]: Wrong~~

File: rageval/metrics/_answer_bleu.py [hunk 4]
```diff
@@ -1,168 +1,165 @@
...
+            bleu_result = bleu.compute(predictions=predictions, references=references)
+            bleu_score = bleu_result['bleu']
+            scores.append(bleu_score)
+
+        return scores
+
+    def compute(
+        self,
+        predictions: List[str],
+        references: List[List[str]],
```
Comment (faneshion) [inline rageval/metrics/_answer_bleu.py:146]: The same as above, using parameters `pred_answers` and `ref_answers`.

File: rageval/metrics/_answer_bleu.py [hunk 5]
```diff
@@ -1,168 +1,165 @@
...
+        bleu_score = bleu_result['bleu']
+        scores = self._compute_single(predictions, references)
+
+        dataset = Dataset.from_dict({
+            "predictions": predictions,
+            "references": references,
+            f"{self.name}_scores": scores
+        })
+
+    def _compute_batch(self, predictions: List[str], references: List[List[str]]) -> List[float]:
```
Comment (faneshion) [inline rageval/metrics/_answer_bleu.py:164]: The same as above, using parameters `pred_answers` and `ref_answers`.

File: rageval/metrics/_answer_chrf.py [hunk 6]
```diff
@@ -1,181 +1,180 @@
...
+        # super()._validate_data(dataset) 
+        if not all(isinstance(answer, str) for answer in predictions):
+            raise ValueError("The type of answers should be a string.")
+        if not all(isinstance(a, List) or not all(isinstance(item, str) for item in a) for a in references):
+            raise ValueError("The type of gt_answers should be a list of strings.")
+
+    def compute(
+        self,
+        predictions: List[str],
+        references: List[List[str]],
```
Comment (faneshion) [inline rageval/metrics/_answer_chrf.py:155]: The same as above, using parameters `pred_answers` and `ref_answers`.

File: rageval/metrics/_answer_claim_recall.py [hunk 7]
```diff
@@ -1,179 +1,178 @@
...
+                "factuality": label,
+            })
+            scores.append(label)
+        # Note that the detail_results can be recorded by logger.info
+        return np.average(scores)
+
+    def _compute_batch(
+        self,
+        predictions: List[str],
+        references: List[List[str]]
```
Comment (faneshion) [inline rageval/metrics/_answer_claim_recall.py:152]: The same as above, using parameters `pred_answers` and `ref_claims`.

File: rageval/metrics/_answer_claim_recall.py [hunk 8]
```diff
@@ -1,179 +1,178 @@
...
+                    "gt_answers": datasets.Value("string")
+                }
+            ),
+            codebase_urls=["https://github.com/princeton-nlp/ALCE"],
+            reference_urls=["https://arxiv.org/abs/2305.14627"]
+        )
+
+    def _compute_one(
+        self,
+        answer: str,
```
Comment (faneshion) [inline rageval/metrics/_answer_claim_recall.py:121]: The same as above, using parameters `pred_answer` and `ref_claims`.

File: rageval/metrics/_answer_disambig_f1.py [hunk 9]
```diff
@@ -1,197 +1,198 @@
...
+        precision = (tp / (tp + fp)) if (tp + fp) > 0 else 1
+        recall = (tp / (tp + fn)) if (tp + fn) > 0 else 1
+
+        if precision + recall == 0:
+            return 0
+        return 2 * (precision * recall) / (precision + recall)
+
+    def _compute_one(
+        self,
+        answer: str,
```
Comment (faneshion) [inline rageval/metrics/_answer_disambig_f1.py:180]: The same as above, using parameters `pred_answer` and `ref_answers`.

File: rageval/metrics/_answer_disambig_f1.py [hunk 10]
```diff
@@ -1,197 +1,198 @@
...
+        for gt_answer in gt_answers:
+            score = self._f1_score(answer, gt_answer)
+            scores.append(score)
+
+        return np.max(scores)
+
+    def _compute_batch(
+        self,
+        predictions: List[str],
+        references: List[List[str]]
```
Comment (faneshion) [inline rageval/metrics/_answer_disambig_f1.py:194]: The same as above, using parameters `pred_answers` and `ref_answers`.

File: rageval/metrics/_answer_edit_distance.py [hunk 11]
```diff
@@ -1,138 +1,137 @@
...
+                    "gt_answers": datasets.Value("string")
+                }
+            ),
+            codebase_urls=[],
+            reference_urls=["https://ieeexplore.ieee.org/abstract/document/10172590"]
+        )
+
+    def _compute_one(
+        self,
+        answer: str,
```
Comment (faneshion) [inline rageval/metrics/_answer_edit_distance.py:103]: The same as above, using parameters `pred_answer` and `ref_answer`.

File: rageval/metrics/_answer_edit_distance.py [hunk 12]
```diff
@@ -1,138 +1,137 @@
...
+                if answer[i - 1] != gt_answer[j - 1]:
+                    dp[i][j] = min(dp[i][j], dp[i - 1][j - 1] + 1)
+                else:
+                    dp[i][j] = min(dp[i][j], dp[i - 1][j - 1])
+
+        return dp[m][n] / m
+
+    def _compute_batch(
+        self,
+        predictions: List[str],
```
Comment (faneshion) [inline rageval/metrics/_answer_edit_distance.py:132]: The same as above, using parameters `pred_answers` and `ref_answers`.

File: rageval/metrics/_answer_exact_match.py [hunk 13]
```diff
@@ -1,132 +1,138 @@
...
+                {
+                    "answers": datasets.Value("string"),
+                    "gt_answers": datasets.Sequence(datasets.Value("string"))
+                }
+            ),
+            codebase_urls=[],
+            reference_urls=["https://arxiv.org/abs/2204.06092"]
+        )
+
+    def _compute_one(self, output: str, short_answers: List[List[str]]) -> float:
```
Comment (faneshion) [inline rageval/metrics/_answer_exact_match.py:114]: The same as above, using parameters `pred_answer` and `ref_answers`.

File: rageval/metrics/_answer_exact_match.py [hunk 14]
```diff
@@ -1,132 +1,138 @@
...
+                if candidate_short_answer in output:
+                    acc.append(True)
+                    break
+            else:
+                acc.append(False)
+        return np.average(acc)
+
+    def _compute_batch(
+        self, 
+        predictions: List[str],
```
Comment (faneshion) [inline rageval/metrics/_answer_exact_match.py:131]: The same as above, using parameters `pred_answers` and `ref_answers`.

File: rageval/metrics/_answer_f1.py [hunk 15]
```diff
@@ -1,160 +1,171 @@
...
+        precision = (tp / (tp + fp)) if (tp + fp) > 0 else 1
+        recall = (tp / (tp + fn)) if (tp + fn) > 0 else 1
+
+        if precision + recall == 0:
+            return 0
+        return 2 * (precision * recall) / (precision + recall)
+
+    def _compute_one(
+        self,
+        answer: str,
```
Comment (faneshion) [inline rageval/metrics/_answer_f1.py:154]: The same as above, using parameters `pred_answer` and `ref_answers`.

File: rageval/metrics/_answer_f1.py [hunk 16]
```diff
@@ -1,160 +1,171 @@
...
+        scores = []
+        for gt_answer in gt_answers:
+            score = self._f1_score(answer, gt_answer)
+            scores.append(score)
+
+        return np.max(scores)
+
+    def _compute_batch(
+        self,
+        predictions: List[str],
```
Comment (faneshion) [inline rageval/metrics/_answer_f1.py:167]: The same as above, using parameters `pred_answers` and `ref_answers`.

File: rageval/metrics/_answer_lcs_ratio.py [hunk 17]
```diff
@@ -1,133 +1,131 @@
...
+                    "gt_answers": datasets.Value("string")
+                }
+            ),
+            codebase_urls=[],
+            reference_urls=["https://ieeexplore.ieee.org/abstract/document/10172590"]
+        )
+
+    def _compute_one(
+        self,
+        answer: str,
```
Comment (faneshion) [inline rageval/metrics/_answer_lcs_ratio.py:104]: The same as above, using parameters `pred_answer` and `ref_answer`.

File: rageval/metrics/_answer_lcs_ratio.py [hunk 18]
```diff
@@ -1,133 +1,131 @@
...
+            for j in range(n):
+                tmp = dp[j + 1]
+                dp[j + 1] = pre + 1 if answer[i] == gt_answer[j] else max(dp[j + 1], dp[j])
+                pre = tmp
+
+        return dp[-1] / m
+
+    def _compute_batch(
+        self,
+        predictions: List[str],
```
Comment (faneshion) [inline rageval/metrics/_answer_lcs_ratio.py:127]: The same as above, using parameters `pred_answers` and `ref_answers`.

File: rageval/metrics/_answer_rouge_correctness.py [hunk 19]
```diff
@@ -113,19 +113,19 @@ def _info(self):
...
-        if not all(isinstance(answer, str) for answer in dataset["answers"]):
+    def _validate_data(self, predictions, references) -> bool:
+        super()._validate_data(predictions, references)
+        if not all(isinstance(answer, str) for answer in predictions):
             raise ValueError("The type of answers should be a string.")
-        if not all(isinstance(a, List) or not all(isinstance(item, str) for item in a) for a in dataset["gt_answers"]):
+        if not all(isinstance(a, List) or not all(isinstance(item, str) for item in a) for a in references):
             raise ValueError("The type of gt_answers should be a list of strings.")
 
     def _compute_one(self, answer: str, gt_answers: List[str]) -> float:
```
Comment (faneshion) [inline rageval/metrics/_answer_rouge_correctness.py:123]: The same as above, using parameters `pred_answer` and `ref_answer`.

File: rageval/metrics/_answer_rouge_correctness.py [hunk 20]
```diff
@@ -113,19 +113,19 @@ def _info(self):
...
         """Evaluate the ROUGE between a single answer and groundtruth answers."""
         score = self.scorer.score_multi(gt_answers, answer)
         return score[self.rouge_type].fmeasure
 
-    def _compute_batch(self, dataset: Dataset) -> list:
+    def _compute_batch(self, predictions, references) -> List[float]:
```
Comment (faneshion) [inline rageval/metrics/_answer_rouge_correctness.py:128]: The same as above, using parameters `pred_answers` and `ref_answers`.

**Review by faneshion — CHANGES_REQUESTED** [review]

**Comment by QianHaosheng** [comment]
I'm wondering why the load diff function isn't working.

**Comment by QianHaosheng** [comment]
Now the input interface of most metrics has been updated, but some new problems have also emerged.

For example:

1. The current test is not comprehensive and needs to be completed, and the correctness of the implementation of each metrics needs to be reconfirmed.
2. The `task` part is not clearly positioned in the project. Now `test_evaluation` cannot work properly due to the interface has been changed, so it is skipped temporarily.
3. The `banchmarks` that calls metrics cannot run normally now, and need to be updated and adapted as soon as possible.
4. No metric actually performs batch calculations currently. We should consider whether to retain the batch to increase parallel calculations or temporarily abandon it.

In addition, there are some formatting issues in the project, which should be fixed in subsequent updates.

**Review by faneshion — APPROVED** [review]

**Pull request merged**